chore: Best Practice-oriented UI refinement (#164)
namehash/ensawards: CONTRIBUTING.md

To add yourself:
1. Add your `AccountId` to the `contributors` collection in [ensawards.org/data/contributors/index.ts](ensawards.org/data/contributors/index.ts) (if this is your first contribution).

> 2. Reference yourself in the contributions array of the entity you are updating. If you update the same entity multiple times, update the `lastUpdated` field for each change with the corresponding timestamp.

Reviewer: ```suggestion
2. Reference yourself in the contributions array of the entity you are updating. If you update the same entity multiple times, update the `lastUpdated` field for each change with the timestamp of your most recent update to that entity.
```

  project: Project; // each protocol is associated with a broader project, which may comprise multiple apps and protocols.
  name: string;
  description: string;
  icon: SvgIcon;
  socials: {
    website: URL;
    twitter: URL;
    ens?: Name;
  };
  /** Relative path from `/data/protocols` to the Open Graph image for the protocol.
   *
   * @optional It's not required to provide an OG image yourself.
   * We have a fallback mechanism in place, so the SEO of Protocol's details page won't be degraded.
   * In fact, we recommend leaving it empty. When your PR with a new {@link Protocol} gets accepted,

>    * the NameHash Labs team will follow it up, providing customized OG images.

Reviewer: ```suggestion
  /** The custom Open Graph image for the protocol. Specified by the relative path from `/data/protocols` to a custom Open Graph image.
   *
   * @optional If not provided, a generic fallback Open Graph image will be used for the protocol. When adding a new protocol we recommend leaving this undefined. The NameHash Labs team will add a custom OG image for your protocol for you.
```

Author: Applied in `data/*/types.ts` and in `CONTRIBUTING.md` for `Apps` and `Protocols`.

   */
  ogImagePath?: string;

  /** Relative path from `/data/protocols` to the Twitter Open Graph image for the protocol.
   *
   * @optional It's not required to provide a Twitter OG image yourself.
   * We have a fallback mechanism in place, so the SEO of Protocol's details page won't be degraded.
   * In fact, we recommend leaving it empty. When your PR with a new {@link Protocol} gets accepted,
   * the NameHash Labs team will follow it up, providing customized OG images.
   */

>   twitterOgImagePath?: string;

Reviewer: Please update the JSDoc for this field to match the general pattern I shared on the other field.

```typescript
export interface App {
  id: string;
  appSlug: AppSlug;
  project: Project; // each app is part of a broader project (which may span multiple apps and protocols).
  name: string;
  description: string;
  type: AppType;

>   icon: (props: React.SVGProps<SVGSVGElement>) => JSX.Element;

Reviewer: The type definition for these icon fields might be intimidating for some people in a way that discourages less technical people from making contributions. For this audience, the code here looks complicated but the idea is actually simple. It's just a svg icon.

A few suggestions:

1. Create a new type alias for `SvgIcon`
2. Use this new type alias in our code wherever we are making reference to relevant icons, including in both our `App` and `Protocol` data models and possibly other places based on your advice.
3. Write a JSDoc comment on the `SvgIcon` type alias explaining in very simple terms what it represents. I think the best way to describe this is just to point to examples in our code and say to do that.

Appreciate your advice on all the details for how best to achieve making this look less technical and less intimidating to data contributors!

Author: @lightwalker-eth 

Prepared the following alias with a JSDoc:

```typescript
/**
 * An SVG icon written as a simple React component.
 *
 * @example
 * For reference on how to implement an `SvgIcon`, see one of the existing icons in the codebase, such as:
 * * `ensawards.org/data/apps/blockscout-explorer/icon.tsx`
 * * `ensawards.org/data/projects/aave/icon.tsx`
 * * `ensawards.org/data/protocols/cork-defi/icon.tsx`
 *
 * You can achieve the desired effect by simply swapping the SVG part from one of the examples with your own.
 */
export type SvgIcon = (props: React.SVGProps<SVGSVGElement>) => JSX.Element;
```

* Focused on simplification:
    * Provided examples as suggested
    * Intentionally avoided speaking about props to reduce complexity
    * A simplified instruction below the examples should put the less technical contributors on the right path
* Additionally, in `CONTRIBUTING.md`, each data model that uses an icon has the following point explicitly mentioned &rarr; *Add an SVG icon as a React component in the created directory (`icon.tsx`). For reference, see [ensawards.org/data/apps/blockscout-explorer/icon.tsx](ensawards.org/data/apps/blockscout-explorer/icon.tsx).*
* This, combined with the JSDoc of the alias, should provide enough information for the contributors

I also want to note that so far, both Spooky and Apoorv were able to add icons without significant issues.

Reviewer: @Y3drk Looks nice!

  name: string;
  description: string;
  type: AppType;
  icon: SvgIcon;
  socials: {
    website: URL;
    twitter: URL;
    ens?: Name;
  };
  /** Relative path from `/data/apps` to the Open Graph image for the app. 
   * 
   * @optional It's not required to provide an OG image yourself. 
   * We have a fallback mechanism in place, so the SEO of App's details page won't be degraded. 
   * In fact, we recommend leaving it empty. When your PR with a new {@link App} gets accepted, 
   * the NameHash Labs team will follow it up, providing customized OG images.
  */

>   ogImagePath?: string;

Reviewer: Please see related comment.

  /** Relative path from `/data/apps` to the Twitter Open Graph image for the app. 
   * 
   * @optional It's not required to provide a Twitter OG image yourself. 
   * We have a fallback mechanism in place, so the SEO of App's details page won't be degraded. 
   * In fact, we recommend leaving it empty. When your PR with a new {@link App} gets accepted, 
   * the NameHash Labs team will follow it up, providing customized OG images.
  */

>   twitterOgImagePath?: string;

Reviewer: Please see related comment.

> {
  type: BestPracticeT;
  id: string;
  bestPracticeSlug: BestPracticeSlug;
  name: string;
  description: string;

>   category: BestPracticeCategory; // each best practice belongs to exactly one category

Reviewer: ```suggestion
  category: BestPracticeCategory; // each best practice belongs to exactly one best practice category
```

export interface BestPracticeCategory {
  id: string;
  categorySlug: BestPracticeCategorySlug;
  name: string;
  description: string;
  status: CategoryStatus;
}
```
4. In your PR describe your reasoning for adding it.

### Suggest a `benchmark update`

1. To suggest a benchmark update for an existing app, modify its `benchmarks` record in the [ensawards.org/data/apps/[app-directory]/benchmarks.ts](ensawards.org/data/apps/rainbow-wallet/benchmarks.ts) file where `[app-directory]` represents the slug of the relevant app. These records define relationships between benchmarks and best practices. Their data model is available at [ensawards.org/data/ens-best-practices/types.ts](ensawards.org/data/ens-best-practices/types.ts).

```typescript
/**
 * Defines relations between {@link BestPracticeSlug} and {@link AppBenchmark}
 * for the related {@link BestPractice}.
 *

>  * @invariant Each {@link BestPractice} must define an explicit entry.

Reviewer: ```suggestion
 * @invariant An explicit key for each `BestPracticeSlug` should be added to this `Record` for each each available {@link BestPractice}. If the related app doesn't have a benchmark completed for the `BestPractice` then the value should explicitly be set to `undefined`. Otherwise, the value should be an `AppBenchmark` describing how the related app performed for the `BestPractice`.
```

 */
export type BestPracticeBenchmarks = Record<BestPracticeSlug, AppBenchmark | undefined>;
```

For reference see [ensawards.org/data/apps/blockscout-explorer/benchmarks.ts](ensawards.org/data/apps/blockscout-explorer/benchmarks.ts) for an example benchmarks record.

2. Make sure to follow the benchmark data model. It is available in the [ensawards.org/data/benchmarks/types.ts](ensawards.org/data/benchmarks/types.ts) file.
```typescript

> export enum BenchmarkResult {

Reviewer: The data model I see documented here is showing UI-level values at a data-model level. I believe I've shared feedback suggesting to separate data-model level representations from UI-level representations.

Please message me in Slack if any questions 👍

    Pass = "Pass",
    PartialPass = "Partial pass",
    Fail = "Fail",
}

export interface AppBenchmark {
  result: BenchmarkResult;

  /** A record of all contributors involved in the addition or maintenance of the benchmark's data.
   *
   * @invariant Multiple {@link Contribution} from the same contributor is not allowed.

>    * When a contributor makes updates to their existing contribution, they should update the `lastUpdated` timestamp.

Reviewer: ```suggestion
   * @invariant Multiple {@link Contribution} from the same contributor are not allowed.
   * When a contributor makes updates to their existing contribution, they should update the `lastUpdated` timestamp from their existing `Contribution`.
```

    Fail = "Fail",
}

> export const BenchmarkStatuses = {

Reviewer: Please see related comments where I suggested this idea can be completely removed.

Author: Resolved via major Benchmark data model refactor (the docs file itself will be updated once the new data model implementation is deemed satisfactory)

  Completed: "Completed",
  Pending: "Pending",
} as const;

> export interface AppBenchmarkAbstract<BenchmarkStatusT extends BenchmarkStatus> {

Reviewer: This is getting too complicated for contributors. We need to put more priority on making it easy for them.

Author: Resolved via major Benchmark data model refactor (the docs file itself will be updated once the new data model implementation is deemed satisfactory)

  /** The best practice being benchmarked */
  bestPractice: BestPracticeApp;

  /** The status of a benchmark */
  status: BenchmarkStatusT;
}

/**
 * Represents a benchmark result for a specific {@link BestPractice} within an {@link App}.
 */

> export interface AppBenchmarkCompleted

Reviewer: Too complicated for contributors.

Author: Resolved via major Benchmark data model refactor (the docs file itself will be updated once the new data model implementation is deemed satisfactory)

 * Defines relations between {@link BestPracticeSlug} and {@link AppBenchmark}
 * for the related {@link BestPractice}.
 *
 * @invariant An explicit key for each `BestPracticeSlug` should be added to this `Record` for each available {@link BestPractice}. 

>  * If the related app doesn't have a benchmark completed for the `BestPractice` then the value should explicitly be set to `undefined`. 

Reviewer: ```suggestion
 * If an app doesn't have a benchmark completed for a `BestPractice` then the benchmark should be explicitly set to `undefined`. 
```

Author: Applied both in `data/ens-best-practices/types.ts` and here.

}
```

> > **NOTE**

Reviewer: We should optimize our data models so that this note is unnecessary.

Author: Resolved via major Benchmark data model refactor (the docs file itself will be updated once the new data model implementation is deemed satisfactory)

2. Make sure to follow the benchmark data model. It is available in the [ensawards.org/data/benchmarks/types.ts](ensawards.org/data/benchmarks/types.ts) file.
```typescript
export const BenchmarkResults = {
  Pass: "passed",
  PartialPass: "partially-passed",
  Fail: "failed",
} as const;

export type BenchmarkResult = (typeof BenchmarkResults)[keyof typeof BenchmarkResults];

export interface AppBenchmark {
  result: BenchmarkResult;

  /** A record of all contributors involved in the addition or maintenance of the benchmark's data.
   *

>    * @invariant Multiple {@link Contribution} from the same contributor are not allowed.

Reviewer: ```suggestion
   * @invariant Multiple {@link Contribution} from the same contributor on the same app benchmark are not allowed.
```

Author: Applied both here and in `data/benchmarks/types.ts`. Did a similar update in `data/procotols/contract-types.ts` to follow the same pattern.

   * When a contributor makes updates to their existing contribution, 

>    * they should update the `lastUpdated` timestamp from their existing `Contribution`.

Reviewer: ```suggestion
   * they should update the `lastUpdated` timestamp of their existing `Contribution`.
```

Author: Applied both here and in `data/benchmarks/types.ts`. Did a similar update in `data/procotols/contract-types.ts` to follow the same pattern.
